### ISSUE TITLE
Does any conflict between amplify_push_notifications_pinpoint and firebase_messaging? 

### ISSUE THREAD

**Issue by cjeremylin** [aws-amplify/amplify-flutter#4237]
### Description

Hi team, 

I noticed that when I install `firebase_messaging` dependency would lead to `Amplify.Notifications.Push.onNotificationReceivedInForeground` can't be triggered when app in foreground. Once I remove it from pubspec.yml and I can receive message in `Amplify.Notifications.Push.onNotificationReceivedInForeground`.  This issue occurred on both physical and emulator of Android API level 29 to 34. 

### Key code snippet
```
 Amplify.Notifications.Push.onNotificationReceivedInForeground
        .listen(foregroundReceiveHandler);
...
Future<void> foregroundReceiveHandler(PushNotificationMessage notification) async {
  log('!foregroundReceiveHandle! $notification');
  //notificationController.messageHandler(notification);
}
```

### main.dart
```
import 'dart:developer';
import 'package:flutter/material.dart';
import 'package:amplify_auth_cognito/amplify_auth_cognito.dart';
import 'package:amplify_flutter/amplify_flutter.dart';
import 'package:amplify_push_notifications_pinpoint/amplify_push_notifications_pinpoint.dart';
import 'amplifyconfiguration.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:amplify_push_flutter/features/notifications/controller/notifications_controller.dart';
//import 'package:firebase_messaging/firebase_messaging.dart';

import 'firebase_options.dart';

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();

  try {
    await _configureFirebase();
    await _configureAmplify();
  } on AmplifyAlreadyConfiguredException {
    debugPrint('Amplify configuration failed.');
  }
  runApp(const MyApp());
}

Future<void> _configureAmplify() async {
  try {
    final authPlugin = AmplifyAuthCognito();
    final pushPlugin = AmplifyPushNotificationsPinpoint();
    pushPlugin.onNotificationReceivedInBackground(
    backgroundNotificationReceivedHandler);
    await Amplify.addPlugins([authPlugin, pushPlugin]);
    log("Pulgin Added");
    await Amplify.configure(amplifyconfig);
    log("Complete confiugre");

    Amplify.Notifications.Push.onTokenReceived.listen((event) {
      log("Push device token: $event");
    });

    Amplify.Notifications.Push.onNotificationReceivedInForeground
        .listen(foregroundReceiveHandler);

    log("END configure AmplifyConfigure");
  } on Exception catch (e) {
    safePrint('An error occurred configuring Amplify: $e');
  }
}

Future<void> backgroundNotificationReceivedHandler(
    PushNotificationMessage notification) async {
  log('backgroundNotificationReceivedHandler $notification');
}

Future<void> foregroundReceiveHandler(
    PushNotificationMessage notification) async {
  log('!foregroundReceiveHandle! $notification');
  //notificationController.messageHandler(notification);
}

Future<void> _configureFirebase() async {
  log("START configure Firebase");
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );
  await notificationController.initialize();
  log("END configure Firebase");
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  @override
  void initState() {
    super.initState();
  }

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

```

We will send push notification via Firebase directly or via Pinpoint in some circumstance. Therefore, we need to include firebase_messaging into dependency due to our business strategy. 

Is any way to achieve that? 

Thanks for looking into the issue. 

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [ ] DataStore
- [X] Notifications (Push)
- [ ] Storage

### Steps to Reproduce

1. Running app and get device token 
2. Put token to Pinpoint console > testing message > Device tokens > send message
3. If firebase_messaging did not include in pubspec.yaml. Mobile phone can receive foreground message 
<img width="1515" alt="without-firebase_messaging" src="https://github.com/aws-amplify/amplify-flutter/assets/5304435/041f6aa8-44ee-4c5f-8bc1-1284755eb48d">

4. If firebase_messaging included in pubspec.yaml and installed. Mobile phone can't receive foreground message. 
<img width="925" alt="with-firebase_messaging" src="https://github.com/aws-amplify/amplify-flutter/assets/5304435/f053cc14-f68c-4ea1-903f-f2d9672684d8">
However, I can see there is a debug message `D/FLTFireMsgReceiver( 9883): broadcast received for message`, and just can't trigger the foreground listener. 

### Screenshots

Please see above in steps

### Platforms

- [X] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.16.3

### Amplify Flutter Version

1.6.1

### Deployment Method

Amplify CLI

**Comment by Jordan-Nelson** [comment]
Hello @cjeremylin - Thanks for opening the issue. I was able to reproduce the experience that you described.

amplify_push_notifications_pinpoint and firebase_messaging both use the Android Firebase SDK. Unfortunately the Android Firebase SDK doesn't support having multiple services set up to receive notifications. More specifically, the SDK requires that the consumer implement FirebaseMessagingService, which both libraries do. The Android Firebase SDK does not support multiple implementations in the same app.

You might be able to work around this by creating your own FirebaseMessagingService implementation which receives notifications and forwards them as needed. You would also have to override the impl from amplify_push_notifications_pinpoint and firebase_messaging which is specified in the app manifest for each lib (see below for location in amplify_push_notifications_pinpoint).

https://github.com/aws-amplify/amplify-flutter/blob/f60e65b7b8412675cd0e0bc7b908b2ae1aaafa4f/packages/notifications/push/amplify_push_notifications/android/src/main/AndroidManifest.xml#L7

Unfortunately there isn't much we can do to support this use case given the existing limitations of the Android Firebase SDK.

Let me know if you have any other questions.

**Comment by Jordan-Nelson** [comment]
@cjeremylin I am going to close this out since the original question has been answered.

**Comment by bitsplash-andy** [comment]
I spent all day trying to find out why the notifications were not working as expected, and it turns out that it is this issue.

Can we add to the docs that these two libraries are incompatible?

It's frustrating, as the Android implementation of push notification handling works far better with the FirebaseMessaging library. We struggled to make notifications appear as we wished using the amplify library, as soon as we added firebase notifications work on Android as expected, but regress on iOS.
